pegasus/tools/scvitools.py

```diff
@@ -109,3 +112,323 @@ def _gen_query_anndata(
+        Keyword in ``data.var``, which refers to a boolean array. If ``None``, all features will be selected.
+    matkey: ``str``, optional, default: ``"raw.X"``
+        Matrix key for the raw count
+    n_jobs : ``int``, optional, default: ``-1``.
+        Number of threads to use. ``-1`` refers to using all physical CPU cores.
+    random_state: ``int``, optional, default: ``0``.
+        Seed for random number generator
+    max_epochs: ``int | None``, optional, default: ``None``.
+        Maximum number of unsupervised training epochs. Defaults to np.min([round((20000 / n_cells) * 400), 400])
+    batch: ``str``, optional, default: ``None``.
+        If only one categorical covariate, the obs key representing batches that should be corrected for, default is ``None``.
+    categorical_covariate_keys: ``List[str]``
+        If multiple categorical covariates, a list of obs keys listing categorical covariates that should be corrected for, default is ``None``.
+    continuous_covariate_keys: ``List[str]``
+        A list of obs keys listing continuous covariates that should be corrected for, default is ``None``.
+    semisupervised_max_epochs: ``int | None``, optional, default: ``None``.
+        Maximum number of semisupervised training epochs. Defaults to np.min([round(np.sqrt(``max_epochs``)), 20])
+    n_samples_per_label : ``int``, optional, default: ``None``.
+        Number of subsamples for each label class to sample per epoch. By default, there is no label subsampling.
+    use_gpu: ``str | int | bool | None``
+        Use default GPU if available (if None or True), or index of GPU to use (if int), or name of GPU (if str, e.g., ‘cuda:0’), or use CPU (if False).
+    arches_params: ``dict``.
+        Hyperparameters for VAE. See https://docs.scvi-tools.org/en/stable/api/reference/scvi.module.VAE.html#scvi.module.VAE for more details
+
+    Returns
+    -------
+    Update ``data.obsm``:
+        * ``data.obsm['X_scVI']``: The embedding calculated by scVI.
+        * ``data.obsm['X_scanVI']``: The embedding calculated by scanVI.
+
+    Examples
+    --------
+    >>> pg.train_scarches_scanvi(data, dir_path="scanvi_model/", label="celltype", matkey="counts", batch="tech", n_samples_per_label=100)
+    """
+    try:
+        import scvi
+    except ImportError as e:
+        import sys
+        logger.error(f"{e}\nscvi-tools needed! Try 'pip install scvi-tools'.")
+        sys.exit(-1)
+
+    logger.info("Start training with scArches method.")
+
+    obs_columns = [label]
+    if batch is not None and batch:
+        obs_columns.append(batch)
+    if categorical_covariate_keys is not None and categorical_covariate_keys:
+        obs_columns.extend(categorical_covariate_keys)
+    if continuous_covariate_keys is not None and continuous_covariate_keys:
+        obs_columns.extend(continuous_covariate_keys)
+
+    adata = _gen_anndata(data, features, obs_columns, matkey) # gen AnnData
+
+    if n_jobs == -1:
+        import multiprocessing
```

Comment:
Please use pg.utils.eff_n_jobs (see `nmf.py` as an example), which set physical cores when possible.

Reply:
done in new commit